src/fmpz_mat/is_spd.c

```diff
@@ -0,0 +1,49 @@
+/*
+    Copyright (C) 2023 Jean Kieffer
+
+    This file is part of FLINT.
+
+    FLINT is free software: you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License (LGPL) as published
+    by the Free Software Foundation; either version 2.1 of the License, or
+    (at your option) any later version.  See <http://www.gnu.org/licenses/>.
+*/
+
+#include "fmpz.h"
+#include "fmpz_mat.h"
+
+int
+fmpz_mat_is_spd(const fmpz_mat_t A)
+{
+    slong d = fmpz_mat_nrows(A);
+    fmpz_mat_t tp, w;
+    fmpz_t det;
+    slong k;
+    int res = 1;
+
+    if (fmpz_mat_ncols(A) != d)
+    {
+        return 0;
+    }
+
+    fmpz_mat_init(tp, d, d);
+    fmpz_init(det);
+
+    fmpz_mat_transpose(tp, A);
+    if (!fmpz_mat_equal(tp, A))
+    {
+        res = 0;
+    }
+
+    for (k = 1; (k <= d) && res; k++)
+    {
+        fmpz_mat_window_init(w, A, 0, 0, k, k);
+        fmpz_mat_det(det, w);
```

Comment:
Can you think of a more efficient way to check symmetric positive definiteness?

If this algorithm is used, it would make sense to add an ``fmpz_mat_sgn_det`` function. This function could be trivial for now, but it could later be optimized by incorporating numerical checks.

Anyway, computing ``d`` determinants looks slow. Isn't a better method here to compute an LDL decomposition and check positivity of D, or something along those lines?

Reply:
This should be better after https://github.com/flintlib/flint/pull/1524/commits/5840f9adb5cca6164d191d4678a1648e6bc39b1d. I'm using a numerical LDL decomposition first to confirm that the matrix is positive definite, and if that fails I used the rule of signs on the characteristic polynomial. It's probably still not the fastest possible code though.

Reply:
I updated the documentation in https://github.com/flintlib/flint/pull/1524/commits/3ed3ec3ff6c2d5cca249ea446fb01859456447b3